src/DataTypes/ObjectUtils.cpp

```diff
@@ -698,4 +698,33 @@ void finalizeObjectColumns(MutableColumns & columns)
             column_object->finalize();
 }
 
+Field FieldVisitorReplaceScalars::operator()(const Array & x) const
+{
+    if (num_dimensions_to_keep == 0)
+        return replacement;
+
+    const size_t size = x.size();
+    Array res(size);
+    for (size_t i = 0; i < size; ++i)
+        res[i] = applyVisitor(FieldVisitorReplaceScalars(replacement, num_dimensions_to_keep - 1), x[i]);
+    return res;
+}
+
+size_t FieldVisitorToNumberOfDimensions::operator()(const Array & x) const
+{
+    const size_t size = x.size();
+    std::optional<size_t> dimensions;
+
+    for (size_t i = 0; i < size; ++i)
+    {
+        size_t current_dimensions = applyVisitor(*this, x[i]);
+        if (!dimensions)
+            dimensions = current_dimensions;
+        else
+            dimensions = std::max(*dimensions, current_dimensions);
+    }
+
+    return 1 + dimensions.value_or(0);
+}
```

Comment:
as I understand using optional is solely to avoid calling max on the first iteration? not sure it worth this complication, besides optional itself adds overhead...
```
size_t FieldVisitorToNumberOfDimensions::operator()(const Array & x) const
{
    const size_t size = x.size();
    size_t dimensions = 0;

    for (size_t i = 0; i < size; ++i)
        dimensions = std::max(dimensions, applyVisitor(*this, x[i]));

    return 1 + dimensions;
}
```